### ISSUE TITLE
Labels in call stack for breakpoints, variables and threads are misaligned and unreadable

### ISSUE THREAD

**Issue by keithrussell42** [daveleroy/SublimeDebugger#223]
I have been struggling with this issue for some time now and can't find a reasonable solution. My environment(s):

* Sublime Text 4 Build 4143 (and happened with previous ST4 builds as well)
* Happens on Windows (10 and 11) and Ubuntu 20.04
* I see this when debugging Go (Delve) and Rust (lldb cargo). I haven't tried anything else.
* Font size is 11 on Linux, 10 on Windows.
* Debugger package v0.10.1

I have tried to reduce the UI scale in the debugger settings. This sometimes fixes it for specific instances but another debug session would have some misaligned labels. I've even tried to use my preferred font but that didn't help either. 

This is 100% reproducible for me. I'm quite willing to play with the code if there are suggestions where to look. 

![image](https://github.com/daveleroy/SublimeDebugger/assets/35411268/78bca241-906d-400a-8f93-05d7e266689b)

**Comment by daveleroy** [comment]
You could try playing around with the value here https://github.com/daveleroy/SublimeDebugger/blob/7c88fa6402e154f41e4db1075bd336bc539b035f/modules/ui/style.py#L100

**Comment by daveleroy** [comment]
Not currently on package control but you can adjust this with the `internal_font_scale` setting 

https://github.com/daveleroy/SublimeDebugger/blob/805614b704da4f61f11fefcc13a833de25cad229/Debugger.sublime-settings#L12

**Comment by WZ-Tong** [comment]
> I have been struggling with this issue for some time now and can't find a reasonable solution. My environment(s):
> 
> * Sublime Text 4 Build 4143 (and happened with previous ST4 builds as well)
> * Happens on Windows (10 and 11) and Ubuntu 20.04
> * I see this when debugging Go (Delve) and Rust (lldb cargo). I haven't tried anything else.
> * Font size is 11 on Linux, 10 on Windows.
> * Debugger package v0.10.1
> 
> I have tried to reduce the UI scale in the debugger settings. This sometimes fixes it for specific instances but another debug session would have some misaligned labels. I've even tried to use my preferred font but that didn't help either.
> 
> This is 100% reproducible for me. I'm quite willing to play with the code if there are suggestions where to look.
> 
> ![image](https://private-user-images.githubusercontent.com/35411268/239434117-78bca241-906d-400a-8f93-05d7e266689b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzMDU1MTIsIm5iZiI6MTcxNjMwNTIxMiwicGF0aCI6Ii8zNTQxMTI2OC8yMzk0MzQxMTctNzhiY2EyNDEtOTA2ZC00MDBhLThmOTMtMDVkN2UyNjY2ODliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIxVDE1MjY1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFlNzVlZmI0OGVmYTllNzg4ZjliNjFiYTExMjMxZTY4ZmJmYzg2YjU5ODM4ZDM5ZTVmNDA1OTcxNTQzNzk5N2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hb05Qas2mjQ3qTVOAbVott44RDwXc9bN7sVQDE2Mc_I)

Hello kei, I encountered the same as you, have you already tried out a set of value that displays correctly? I found my "Variable" panel been "Squeesed" into "Next page"

**Comment by daveleroy** [comment]
Are you using `internal_font_scale` or `ui_scale`?

**Comment by WZ-Tong** [comment]
> @WZ-Tong Are you using `internal_font_scale` or `ui_scale`?

Yes, but I found it's useless to "un-squeeze" the `Variable` panel even if I use `internal_font_scale=0.75`, `ui_scale=10` (Cannot be reduced while keeps readability).
Under this setting, the UI looks like:
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/4ba6ccc7-dcbb-4212-a461-38f7a4582e88)
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/267b4ecc-c3ff-45cb-8a38-aacef9d481d3)
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/2bae3533-0bbd-4b6d-92ca-e50ebaa35f2d)

**Comment by WZ-Tong** [comment]
> @WZ-Tong Are you using `internal_font_scale` or `ui_scale`?

But the question kei met on this issue, just changing the `internal_font_scale` to 0.97 may help, but the `Variable` panel is still squeezed out.
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/f16aebd1-440a-420c-8cb4-823fe9bb9b65)

**Comment by WZ-Tong** [comment]
> @WZ-Tong Are you using `internal_font_scale` or `ui_scale`?

I think this is caused by rust's backtrace name (it contains the full module path, and it's often quite long).

And I believe the `Variable` panel is also "just a little bit bigger, so cannot fit in a line", since I can just fix kei's problem by shirinking `internal_font_scale` just a little bit.

**Comment by WZ-Tong** [comment]
@daveleroy So, could I ask politely that if it's easy to fix because it's really a problem for rust user to use the debugger. I would appreciate it if some small improvements can be applied, but it's hard for me to create a PR myself, since I'm not familiar with related things.

**Comment by daveleroy** [comment]
I don't think the issue is with rust here many adapters have content that is too large for the space.

Everything rendered in the debugger has to be sized and clipped to fit inside the available space and that isn't happening for some users but its not clear why the calculations aren't working. I believe setting `internal_font_scale` fixes the issue for most people which changes the size of the text. 

You appear to have some other issue where the panel (not the text/images) are too large or at least thats what it appears since the text/images is smaller than the panel when setting `internal_font_scale`

**Comment by WZ-Tong** [comment]
Yes, so I think my problem is more similar to the closed one #253. I'm trying to figure out where is the key point

**Comment by WZ-Tong** [comment]
https://github.com/daveleroy/SublimeDebugger/blob/cb9971093bdb255d943c6d0ae5caff3641130f2f/modules/callstack.py#L63

In my case, I can change this code to fix my problem:
```python
with TabbedViewContainer(width_scale=0.5, width_additional=-32, width_additional_dip=-32):
```
But I don't know that if there are any side effects.

After applying this, my ui appears like:
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/c226eb5f-7d09-45d3-a615-1c6f235fe264)
There are some empty spaces, so I don't know why the original panel is "wrapped" into the second line

Whatever, it's now useable and useful. Thanks for providing this plugin!

**Comment by daveleroy** [comment]
@WZ-Tong That shouldn't cause any issues is basically making the panels smaller than the window size. The panels are intended to be sized to exactly fit the window and that appears to be calculated wrong. The current calculation is basically the window size - 30 dip since each phantom has some padding around them that appears to be 10 dip (5 on each side)

If you are willing to try something that might get everything to line up correctly you could remove all your changes and try making a change here.

https://github.com/daveleroy/SublimeDebugger/blob/cb9971093bdb255d943c6d0ae5caff3641130f2f/modules/ui/layout.py#L149

into 

``` python
def from_dip(self, dip: float) -> float:
	return (dip / self.em_width) * self.internal_font_scale // (maybe try adding - 0.001) or something as well
```

And then using the `internal_font_scale` setting and see if just setting that causes everything to line up perfectly and not wrap the panel. Including the end of lines here with the end of the panel. That change would only change how the 10 dip is calculated

 ![Screenshot 2024-05-24 at 12 28 41 AM](https://github.com/daveleroy/SublimeDebugger/assets/2889367/edc31974-aab3-4c44-9ce5-e49a97a102c3)

**Comment by daveleroy** [comment]
Yeah thats too far off to be the issue with the panel width calculation. 

I think the solution here for now is to change the `width_additional_dip` which adds the 30dip for the space sublime adds between the phantoms to something larger.

If you remove all the changes you made besides your `internal_font_scale` setting what value for `width_additional_dip` works here?

https://github.com/daveleroy/SublimeDebugger/blob/cb9971093bdb255d943c6d0ae5caff3641130f2f/modules/callstack.py#L63

If you don't have to fudge `width_additional_dip` too much I'll make a change and commit it.

Leave `width_additional` at -30 thats for the size of the breakpoints panel

**Comment by WZ-Tong** [comment]
Sorry, but seems it's my fault that I forget to remove `width_additional=-32` in the prev test.
I get a fresh install using package control `remove` + `install`, here are some data (All of these are under `internal_font_scale=0.97`):

Failed:
```python
width_additional=-30, width_additional_dip=-35
width_additional=-30, width_additional_dip=-40
width_additional=-31, width_additional_dip=-30
```

Works:
```python
width_additional=-32, width_additional_dip=-30
```

**Comment by WZ-Tong** [comment]
After applying `width_additional=-32, width_additional_dip=-30`, the UI works under `internal_font_scale=1 (unmodified)` perfectly

**Comment by daveleroy** [comment]
That unfortunately causes a pretty big gap so I'll add an additional setting to change this because I can't seem to figure out what is why the calculations are not working here.

From your screenshot if you make the window really small does the gap on the very right stay the same size?

![Screenshot 2024-05-25 at 1 13 48 AM](https://github.com/daveleroy/SublimeDebugger/assets/2889367/6c49ed34-6fac-40f8-bc43-fc129c08573a)

If it stays the same size I will make a fixed `internal_width_modifer` setting that adds to the adds to `width_additional` 

If it changes size I will make an `internal_width_scale` setting that scales the width that the ui thinks is available.

**Comment by WZ-Tong** [comment]
Take `internal_font_scale=1, width_additional=-32, width_additional_dip=-30` for example

If the window is smaller than 948*Height pixel, the variable panel is sqeezed out (like issue #253)
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/39a947ce-0c78-43bc-90a8-ef31522b1980)
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/638b950e-dab7-4640-9c96-bb5167b6802b)

But I noticed that if the window is 940~947 pixel wide, and I'm dragging it from smaller (e.g 930 pixel) to bigger, the variable panel pops up, and after maybe re-calc its size, it is squeezed out again. This maybe the threashhold.
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/0770463a-0c21-4ff6-9f94-a0ba5f8d0e1d)
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/eee1f495-6042-4106-a835-0cca8b44fd86)

By calc the pixel from the right side of the panel, to the edge of sublime text window, I got the following:
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/c382ded4-d40c-45c4-9a8e-f08c959500d8)
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/8f229cec-ddcf-4ceb-89e4-46eab5f369e1)
![image](https://github.com/daveleroy/SublimeDebugger/assets/77199564/78de5bf1-1a83-422a-9480-afe5f0dd583d)

It stays in the same size

**Comment by daveleroy** [comment]
I'll go ahead and add an `internal_width_modifer` setting. 

Thanks for going through all this stuff, grabbing screenshots and testing all the cases its been a big help!

**Comment by daveleroy** [comment]
The changes are on the main branch.

 I'm a little confused though because when I use the new setting and make the panel too big it just adds a scroll bar not sure why yours is even wrapping instead of doing the same.

**Comment by WZ-Tong** [comment]
Sorry to be late. It works perfectly now! 

I don't know why either, but I have to mention that I'm using Sublime Text Dev build 4175, which maybe cause some differences?

BTW, the version is different from package control, maybe you should push to package control again?

**Comment by waelbachta** [comment]
Hello, 

I am working under a Mac OS machine, and I cannot resize or displace Variables, the Callstack, and the Console tabs. Is this behavior expected?

Thanks a lot for your help.

Wael

**Comment by WZ-Tong** [comment]
I'm running st4 latest build on windows, the panel will be automatically resized after resizing st4 window, but it cannot be manually resized by dragging the boundary. I think it's the expected behavior (not officially, I'm just a user)

**Comment by waelbachta** [comment]
@WZ-Tong I am using the build 4192 of ST. My installation behaves like yours but I was expecting to do whatever I want by dragging the boundaries.

**Comment by waelbachta** [comment]
Ok. Thank you for your time.
